src/pke/examples/iterative-ckks-bootstrapping.cpp

```diff
@@ -0,0 +1,193 @@
+precision as a parameter to EvalBootstrap with multiple iterations. With 2 iterations, you can achieve
+double the precision of a single bootstrapping.
+
+* Source: Bae Y., Cheon J., Cho W., Kim J., and Kim T. META-BTS: Bootstrapping Precision
+* Beyond the Limit. Cryptology ePrint Archive, Report
+* 2022/1167. (https://eprint.iacr.org/2022/1167.pdf)
+
+*/
+
+#define PROFILE
+
+#include "openfhe.h"
+
+using namespace lbcrypto;
+
+void IterativeBootstrapExample();
+
+int main(int argc, char* argv[]) {
+    // We run the example with 8 slots and ring dimension 4096.
+    IterativeBootstrapExample();
+}
+
+// CalculateApproximationError() calculates the precision number (or approximation error).
+// The higher the precision, the less the error.
+double CalculateApproximationError(const std::vector<std::complex<double>>& result,
+                                   const std::vector<std::complex<double>>& expectedResult) {
+    if (result.size() != expectedResult.size())
+        OPENFHE_THROW(config_error, "Cannot compare vectors with different numbers of elements");
+
+    // using the Euclidean norm
+    double maxError = 0;
+    for (size_t i = 0; i < result.size(); ++i) {
+        double error = std::abs(result[i].real() - expectedResult[i].real());
+        if (maxError < error)
+            maxError = error;
+    }
+
+    return std::abs(std::log2(maxError));
+}
+
+void IterativeBootstrapExample() {
+    // Step 1: Set CryptoContext
+    CCParams<CryptoContextCKKSRNS> parameters;
+    SecretKeyDist secretKeyDist = UNIFORM_TERNARY;
+    parameters.SetSecretKeyDist(secretKeyDist);
+    parameters.SetSecurityLevel(HEStd_NotSet);
+    parameters.SetRingDim(1 << 12);
+
+    /*  A3) Key switching parameters.
+    * By default, we use HYBRID key switching with a digit size of 3.
+    * Choosing a larger digit size can reduce complexity, but the size of keys will increase.
+    * Note that you can leave these lines of code out completely, since these are the default values.
+    */
+    parameters.SetNumLargeDigits(3);
+    parameters.SetKeySwitchTechnique(HYBRID);
+
+#if NATIVEINT == 128
+    // Currently, only FIXEDMANUAL and FIXEDAUTO modes are supported for 128-bit CKKS bootstrapping.
+    ScalingTechnique rescaleTech = FIXEDAUTO;
+    usint dcrtBits               = 78;
+    usint firstMod               = 89;
+#else
+    // All modes are supported for 64-bit CKKS bootstrapping.
+    ScalingTechnique rescaleTech = FLEXIBLEAUTO;
+    usint dcrtBits               = 59;
+    usint firstMod               = 60;
+#endif
+
+    parameters.SetScalingModSize(dcrtBits);
+    parameters.SetScalingTechnique(rescaleTech);
+    parameters.SetFirstModSize(firstMod);
+
+    std::vector<uint32_t> levelBudget = {3, 3};
+    uint32_t approxBootstrapDepth     = 9;
+    std::vector<uint32_t> bsgsDim     = {0, 0};
+
+    uint32_t levelsUsedBeforeBootstrap = 10;
+    usint depth =
```

Comment:
This estimate appears to be inaccurate. When I check the output after two iterations, I get 9 available levels instead of expected 10.

Reply:
Confirmed and changed to 9.

Reply:
I don't think this is the right fix. The meta-bts bootstrapping requires one extra level on top of regular ckks bootstrapping. I did not check it. But the depth of bootstrapping ciphertext is probably now 8 instead of the expected 9.

Reply:
For some reason, I'm getting 10 available levels now. I updated approxBootstrapDepth to be 8 + (numIterations + 1).

Reply:
Works correctly now.